logic/src/commonMain/kotlin/com/wire/kalium/logic/data/conversation/MLSConversationRepository.kt

```diff
@@ -514,6 +520,28 @@ internal class MLSConversationDataSource(
             wrapMLSRequest { mlsClient.isGroupVerified(idMapper.toCryptoModel(groupID)) }
         }.map { it.toModel() }
 
+    override suspend fun rotateKeysAndMigrateConversations(
+        clientId: ClientId,
+        e2eiClient: E2EIClient,
+        certificateChain: String
+    ) = mlsClientProvider.getMLSClient().flatMap { mlsClient ->
+        wrapMLSRequest {
+            mlsClient.e2eiRotateAll(e2eiClient, certificateChain, 10U)
+        }.map { rotateBundle ->
+            // todo: make below API calls atomic when the backend does it in one request
+            kaliumLogger.w("drop old key packages after conversations migration")
+            keyPackageRepository.deleteKeyPackages(clientId, rotateBundle.keyPackageRefsToRemove)
+
+            kaliumLogger.w("upload new key packages including x509 certificate")
+            keyPackageRepository.uploadKeyPackages(clientId, rotateBundle.newKeyPackages)
+
+            kaliumLogger.w("send migration commits after key rotations")
+            rotateBundle.commits.forEach {
+                sendCommitBundle(GroupID(it.key), it.value)
+            }
+        }
+    }
+
```

Comment:
Can we cover it with test ?

Reply:
Thanks for the comment, already tests were WIP, just pushed them.